src/iminuit/qtwidget.py

```diff
@@ -0,0 +1,413 @@
+                    transform=self.fig.get_axes()[-1].transAxes,
+                    fontsize="x-large",
+                    ha="right",
+                )
+
+        def fit(self):
+            if self.algo_choice.currentText() == "Migrad":
+                minuit.migrad()
+            elif self.algo_choice.currentText() == "Scipy":
+                minuit.scipy()
+            elif self.algo_choice.currentText() == "Simplex":
+                minuit.simplex()
+                return False
+            else:
+                assert False  # pragma: no cover, should never happen
+            return True
+
+        def on_parameter_change(self, from_fit=False, report_success=False):
+            if any(x.fit.isChecked() for x in self.parameters):
+                saved = minuit.fixed[:]
+                for i, x in enumerate(self.parameters):
+                    minuit.fixed[i] = not x.fit.isChecked()
+                from_fit = True
+                report_success = self.do_fit(plot=False)
+                self.results_text.clear()
+                self.results_text.setHtml(minuit._repr_html_())
+                minuit.fixed = saved
+            elif from_fit:
+                self.results_text.clear()
+                self.results_text.setHtml(minuit._repr_html_())
+            else:
+                self.results_text.clear()
+
+            for ax in self.fig.get_axes():
+                ax.clear()
+            self.plot_with_frame(from_fit, report_success)
+            self.canvas.draw_idle()
+
+        def do_fit(self, plot=True):
+            report_success = self.fit()
+            for i, x in enumerate(self.parameters):
+                x.reset(val=minuit.values[i])
+            if not plot:
+                return report_success
+            self.on_parameter_change(from_fit=True, report_success=report_success)
+
+        def on_update_button_clicked(self):
+            for x in self.parameters:
+                x.slider.setTracking(self.update_button.isChecked())
+
+        def on_reset_button_clicked(self):
+            minuit.reset()
+            minuit.values = original_values
+            minuit.limits = original_limits
+            for i, x in enumerate(self.parameters):
+                x.reset(val=minuit.values[i], limits=original_limits[i])
+            self.on_parameter_change()
+
+    if qt_exec:
+        app = QtWidgets.QApplication.instance()
+        if app is None:
+            app = QtWidgets.QApplication([])
+        main_window = MainWindow()
+        main_window.show()
+        app.exec()
+    else:
+        return MainWindow()
+
+
+def _make_finite(x: float) -> float:
+    sign = -1 if x < 0 else 1
+    if abs(x) == np.inf:
+        return sign * sys.float_info.max
+    return x
+
+
+def _guess_initial_step(val: float, vmin: float, vmax: float) -> float:
+    if np.isfinite(vmin) and np.isfinite(vmax):
+        return 1e-2 * (vmax - vmin)
+    return 1e-2
```

Comment:
This look like they were copied from elsewhere. Import them from the ipywidget or from utils, wherever they came from.

Reply:
They are from ipywidget, but I don't want to import them from there, since then I would also be importing the ipywidgets package or is there a way to prevent that I am not aware of? I could move these functions into utils and both ipywidget and qtwidget then import them from there?

Reply:
Yes, please move them to utils then.

Reply:
They are in util now, but I renamed them, because there was already one with the name _guess_initial_step in there. I added a _widget in front of the names.
https://github.com/scikit-hep/iminuit/blob/a50b5e52cff4a492f1a24ebf1c65e38ac60c4f34/src/iminuit/util.py#L1689-L1699

Reply:
The widget prefix for guess_initial_step is fine, but please keep the simpler name _make_finite, because that function is general and has no tight link to widgets.

Reply:
Ok, I changed it back to _make_finite.